CARDS-1716: As a patient user, I can authenticate without a token

acrowthe opened the pull request:
Known improvements:
- Add restricted tokens so tokenless authentication users have less data access (patient information form, visit information form, submitted surveys)

Testing:
- Create patients with a DOB and MRN or HC
- Create visits with various values for surveys, visit time (future or past) and status ("cancelled" or anything else)

Only visits with a survey answer, a visit time in the future and a status other than "cancelled" should be shown.

On the Proms homepage, enter a DOB and MRN or HC. If this is not a valid patient or if the patient doesn't have any valid visits, an "invalid credentials" error should be shown. There should be no way for a user to differentiate between an invalid patient and valid patient with no valid visits.
If the patient is valid and they have 1 valid upcoming visit, the login process should proceed as before (terms of use if required, then form list and submission).
If the patient is valid and they have >1 valid upcoming visit, they should be presented with a list of "survey" answers for each upcoming visit. They should then be able to select one of these, and then proceed with completing that visit.

acrowthe committed:
CARDS-1716: As a patient user, I can authenticate without a token
Known improvements: Add restricted tokens so tokenless authentication users have less data access

sashaandjic commented:
Tried visits with `pending` and `planned` status. I was not able to login as a patient with and without token.

https://user-images.githubusercontent.com/23243244/164886497-ee988d3e-fb48-4692-99b7-e08135d53930.mp4

https://user-images.githubusercontent.com/23243244/164886502-68de1ed9-d4f2-4e25-b26e-653f9e3cd13d.mp4

https://user-images.githubusercontent.com/23243244/164886505-a20998ef-2bf6-4e13-bc35-3fb8b3f6270d.mp4

https://user-images.githubusercontent.com/23243244/164886512-4efafd75-dd09-459f-88f9-c4501432a1a9.mp4

acrowthe commented:
@sashaandjic I believe I've got the health card number issue reproduced. Looks like the login form is sanitizing the HC number by removing all characters but capital letters and numbers, so `1234567890 AA` is being sanitized to `1234567890AA`. When filling out the patient information forms, make sure enter `123456789AA` or similar to avoid the sanitized login field from being different than the un-sanitized form field.

Looking at the other issues you are having, I'm wondering if those are results of token cookies from previous login attempts. It's working reliably for me if I navigate to `localhost:8080/system/sling/logout` before every login attempt, but if I don't manually log out then I'm seeing very similar errors to what you are having.

Token authentication:
https://user-images.githubusercontent.com/20056751/164915555-b74bcee9-234c-4a6b-829c-35cf933016e3.mp4

Tokenless authentication:
https://user-images.githubusercontent.com/20056751/164915560-6f473109-91a2-469f-9bbc-9aa931aa90e7.mp4

acrowthe committed:
CARDS-1716: As a patient user, I can authenticate without a token
- Change `planned` check into a `cancelled` check to match valiud statuses

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/ToUDialog.jsx:
```diff
@@ -152,7 +167,7 @@ function ToUDialog(props) {
         <Alert severity="error">
           <AlertTitle>An error occurred</AlertTitle>
-          {error}
+          {error ? error : null}
```
This is already in a block conditioned by `error`, so the else part of the condition should never happen.

marta- replied:
We already check if error is truthy 3 lines above, this change is unnecessary.
```suggestion
          {error}
```

acrowthe committed:
CARDS-1716: As a patient user, I can authenticate without a token
- Temporarily remove patient access restriction from audit_sex
  - This restriction exposes an underlying issue with patient user question
    restrictions that leads to white screens

sdumitriu committed:
[misc] Fixed NPE when serializing an answer for an inaccessible question

sdumitriu committed:
Revert "CARDS-1716: As a patient user, I can authenticate without a token"

This reverts commit 56cac3c175c32132e87af7caf222acfc3aff70f4.

The underlying issue has been fixed.

sdumitriu committed:
CARDS-1716: As a patient user, I can authenticate without a token
Also restrict access to the answers of question AUDITC/audit_sex, not just the question itself

sashaandjic commented:
It would be good if we use clinic name rather than SurveyID in this diaolg:
![image](https://user-images.githubusercontent.com/23243244/164950708-b454dbd5-b3ce-4cfe-b38f-dfa352068bac.png)

sashaandjic commented:
The welcome screen after patient successfully login with DOB +MRN (or HC):
![image](https://user-images.githubusercontent.com/23243244/164950722-9c597bf6-384d-49c6-abc1-c4d48299c3fe.png)

sashaandjic commented:
Once patient select one of many visits, the patient is stuck with that visit only.
For example, patient has 1 visit with Cardio and 1 with Mood clinic. Once patient select e.g. Cardio clinic, there is no way for patient to go back and select Mood clinic, even after submitting Cardio surveys.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/PatientIdentification.jsx:
```diff
@@ -213,6 +229,42 @@ function PatientIdentification(props) {
+      <DialogContent>
+        <MaterialTable
+          title="Select an upcoming visit"
+          columns={[{title: "Clinic", field: "survey"}]}
```
> It would be good if we use clinic name rather than SurveyID in this diaolg

```suggestion
          columns={[{title: "Clinic", field: "location"}]}
```

marta- replied:
Issue addressed in 51c55ac.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -435,6 +435,7 @@ function QuestionnaireSet(props) {
   const greet = (name) => {
     let hourOfDay = (new Date()).getHours();
     let timeOfDay = hourOfDay < 12 ? "morning" : hourOfDay < 18 ? "afternoon" : "evening";
+    // TODO: Remove comma with tokenless authentication
     return `Good ${timeOfDay}, ${name}`;
```
```suggestion
    return `Good ${timeOfDay}` + (name ? `, ${name}` : '')
```

marta- replied:
Done in 4a6eed6.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -456,6 +457,7 @@ function QuestionnaireSet(props) {
   }
 
   let appointmentAlert = () => {
+    // TODO: Remove with tokenless authentication
```
We don't need to remove it. Instead, the patient should have access only to the `location`, and the the message will read "Upcoming appointment at <location>", which is both informative and not disclosing PII".
```suggestion
```

marta- replied:
Done in 19969cb.

marta- commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -67,7 +67,7 @@
       "scripts": [
         // Access rights for the special "patient" user: only allow access to the session subject data, excluding certain clinician-only sections/questions
-        "create user patient \n set ACL for patient \n     allow jcr:read on /Questionnaires,/SubjectTypes \n     deny jcr:all on /Questionnaires restriction(rep:itemNames,audit_intro,audit_results,eq5d_results,gad7_intro,gad7_results,phq9_intro,phq9_results) \n     allow jcr:read,jcr:write,jcr:nodeTypeManagement,jcr:versionManagement on /Forms restriction(cards:sessionSubject) \n     deny jcr:re ...
+        "create user patient \n set ACL for patient \n     allow jcr:read on /Questionnaires,/SubjectTypes \n     deny jcr:all on /Questionnaires restriction(rep:itemNames,audit_intro,audit_results,audit_sex,eq5d_results,gad7_intro,gad7_results,phq9_intro,phq9_results) \n     allow jcr:read,jcr:write,jcr:nodeTypeManagement,jcr:versionManagement on /Forms restriction(cards:sessionSubject) \n     d ...
```
> For non-token logins we should not display the following on the welcome screen:
> 
> * patient name (display the greeting without the name)
> * upcoming visit information box

I would leave the `patient` user as is and introduce a `guest-patient` used when the patient logs in through the landing page rather than using the url with an auth token sent by email. The `guest-patient` would have their access denied to more data than `patient`, including `Patient information` entirely and everything from `Visit information` except the `location`.

marta- replied:
`guest-patient` introduced in c5eca16. To do: restrict access to Patient information and Visit information.

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/ValidateCredentialsServlet.java:
```diff
@@ -76,60 +88,195 @@ public class ValidateCredentialsServlet extends SlingAllMethodsServlet
+            Calendar tokenExpiryDate = Calendar.getInstance();
+            tokenExpiryDate.add(Calendar.HOUR_OF_DAY, 1);
+            final String token = this.tokenManager.create(
+                "patient",
```
Use `guest-patient` instead

marta- replied:
Done in c5eca16.

sashaandjic commented:
Another issue: Upon login as a patient I am getting submitted visits in the dialog:
![image](https://user-images.githubusercontent.com/23243244/164952403-dde0ee76-8b4a-4185-98e2-027999efd0a2.png)

sashaandjic commented:
We should allow blanks in HC (after all we do suggest that there will be blanks):
![image](https://user-images.githubusercontent.com/23243244/164952556-54ea4519-c448-49dc-90ed-f2c97b71113f.png)

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/ValidateCredentialsServlet.java:
```diff
@@ -173,6 +320,59 @@ private Node getPatientInformationQuestionnaire(final Session session) throws Re
+        Node statusAnswer = this.formUtils.getAnswer(visitInformationForm, visitStatusQuestion);
+        if (!statusAnswer.hasProperty(VALUE) || "cancelled".equals(statusAnswer.getProperty(VALUE).getString())) {
+            result = false;
+        }
```
> Another issue: Upon login as a patient I am getting submitted visits in the dialog

Add a check that `surveys_submitted` is not `true`

marta- replied:
Done in a84ef2b.

marta- commented:
> We should allow blanks in HC (after all we do suggest that there will be blanks)

We do allow the patient to enter blanks. We take out those blanks before sending the data to the server for patient authentication because the health card numbers we import from Torch don't have blanks in them and we have to match that format.

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
When displaying upcoming appointments, label them by location rather than surveys id

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Exclude visits where the surveys have already been submitted from the upcoming appointment list

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Better support for inaccessible patient name

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Add sign out functionality to the patient UI, to facilitate returning to the homepage for filling out surveys for another visit

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Refactoring: Added a `guest-patient` user for patients who sign in without a token, using the landing page directly rather than following the link in the notification email.
The access to the Patient information and Visit information data is not yet denied for `guest-patient`..

marta- commented:
> Once patient select one of many visits, the patient is stuck with that visit only. For example, patient has 1 visit with Cardio and 1 with Mood clinic. Once patient select e.g. Cardio clinic, there is no way for patient to go back and select Mood clinic, even after submitting Cardio surveys.

Somewhat addressed in ec57673. I added a "Close" button at the bottom of the summary screen, which sends the patient back to the landing page. From there they can sign in again for their next survey. The workflow is not very efficient, but the likelihood that a patient will truly be in this situation is extremely small at this point. We can optimize later.

marta- commented:
> It would be good if we use clinic name rather than SurveyID in this diaolg

Done in 51c55ac.

marta- commented:
> The welcome screen after patient successfully login with DOB +MRN (or HC): For non-token logins we should not display the following on the welcome screen:
> 
> * patient name (display the greeting without the name)
> * upcoming visit information box
> 

Started addressing this in c5eca16 . @sdumitriu please set up the `guest-patient` permissions appropriately.

marta- commented:
> Another issue: Upon login as a patient I am getting submitted visits in the dialog

Fixed in a84ef2b.

sashaandjic commented:
Much better! Like the `Sign out` and `CLOSE` buttons.
The remaining things to fix:
- For non-token logins remove from the welcome screen patient name and upcoming visit information box
- If possible, if there are no surveys to fill out, display the proper message instead of "Invalid credentials"

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Exclude visits with no surveys scheduled from the upcoming appointment list

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Addressing comments - usability improvements:
* Acknowledge successful identification when the patient enters the correct credentials but has no pending surveys to fill out, instead of displaying the generic "Invalid credentials" message
* When there is more than one upcoming visit, display the list within the page rather than in a separate dialog
* Other minor wording, ui, code style adjustments

marta- commented:
> * If possible, if there are no surveys to fill out, display the proper message instead of "Invalid credentials"

Addressed in 89fe69e.

sdumitriu committed:
CARDS-1733: Questionnaire ACL restrictions should support multiple questionnaire paths

sdumitriu committed:
CARDS-1716: As a patient user, I can authenticate without a token
Forbid all access to the Patient Information form for guest-patient

sdumitriu committed:
CARDS-1716: As a patient user, I can authenticate without a token
Don't return the patient's name when validating a tokenless patient

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/PatientIdentification.jsx:
```diff
@@ -169,6 +191,12 @@ function PatientIdentification(props) {
     identify();
   }
 
+  useEffect(() => {
+    if (!visit) return;
+    setVisitList(null);
+    identify();
+  }, [visit]);
```
Why?

marta- replied:
Explained and optimized in 8340d77.

sdumitriu reviewed (approved):
Still left to do: since there are two `identify` calls, the second one with a token, the backend sends the user's name in reply (as expected for a "patient presented a token" authentication), the name is displayed to the user.

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
* Deny access to time, location, provider from Visit information to the `guest-patient`
* As a consequence, do not display appointment information to `guest-patient` after successfully signing in

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/ValidateCredentialsServlet.java:
```diff
@@ -402,24 +402,26 @@ private Node getPatientInformationForm(final Node visitSubject, final Node patie
     }
 
     private void writeSuccess(final SlingHttpServletResponse response, final String sessionSubjectIdentifier,
-        final Node visitSubject, final Session session)
+        final Node visitSubject, final Session session, final boolean includePatientInformation)
```
I don't like hardcoding this in the java code. Is there a way we could get the `Patient information` as the authenticated user (either `patient` or `guest-patient` depending on the path they took to sign in) and return what that user can access instead?

marta- replied:
Nevermind, I misread the changes. All good.

marta- committed:
[misc] Minor code cleanup

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Optimization: Only re-identify upon the change of the `visit` state variable it was updated as a result of the token-less user selecting one of multiple available visits to fill out surveys for.

marta- commented:
> Still left to do: since there are two `identify` calls, the second one with a token, the backend sends the user's name in reply (as expected for a "patient presented a token" authentication), the name is displayed to the user.

This seems to have been solved (or at least covered-up?) by 8340d77.

marta- commented:
> * For non-token logins remove from the welcome screen patient name and upcoming visit information box

Done (in too many commits to list here).

marta- commented:
There's one more workflow improvement that I'd like to make, but it doesn't seem urgent at this time: when signing in as a patient for the first time, after entering the credentials, it would make sense to first pop up the Terms of Use and then once they've accepted show the list of upcoming appointments. That would mean generating a short-lived token on the Patient subject first with the purpose of sorting out the Tems of Use, and after acceptance is confirmed proceed to retrieving upcoming visits.

acrowthe commented:
Logging improvement:
Created a patient with a few invalid visits (submitted, cancelled, and no forms) and a single valid visit. Upon tokenless sign in, I am correctly sent to the one valid visit, but I 2000+ lines written to the error log, mostly access denied on null:
```
25.04.2022 10:30:15.137 *WARN* [[0:0:0:0:0:0:0:1] [1650897014684] GET /Subjects/3db01c81-cc91-48b0-b6e7-d4b030aecd61/433be44f-6ec9-4433-a08e-d379ec03fa75.data.deep.json HTTP/1.1] io.uhndata.cards.dataentry.internal.serialize.AbstractAnswerCopyProcessor Failed to look for the right answer to copy: null
javax.jcr.AccessDeniedException: null
	at org.apache.jackrabbit.oak.jcr.session.NodeImpl$1.perform(NodeImpl.java:194) [org.apache.jackrabbit.oak-jcr:1.40.0]
	at org.apache.jackrabbit.oak.jcr.session.NodeImpl$1.perform(NodeImpl.java:185) [org.apache.jackrabbit.oak-jcr:1.40.0]
	at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.perform(SessionDelegate.java:209) [org.apache.jackrabbit.oak-jcr:1.40.0]
	at org.apache.jackrabbit.oak.jcr.session.ItemImpl.perform(ItemImpl.java:112) [org.apache.jackrabbit.oak-jcr:1.40.0]
	at org.apache.jackrabbit.oak.jcr.session.NodeImpl.getParent(NodeImpl.java:185) [org.apache.jackrabbit.oak-jcr:1.40.0]
	at io.uhndata.cards.dataentry.internal.serialize.FormAnswerCopyProcessor.findForm(FormAnswerCopyProcessor.java:102) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
	at io.uhndata.cards.dataentry.internal.serialize.AbstractAnswerCopyProcessor.getAnswer(AbstractAnswerCopyProcessor.java:151) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
	at io.uhndata.cards.dataentry.internal.serialize.AbstractAnswerCopyProcessor.copyAnswers(AbstractAnswerCopyProcessor.java:128) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
	at io.uhndata.cards.dataentry.internal.serialize.AbstractAnswerCopyProcessor.leave(AbstractAnswerCopyProcessor.java:102) [io.uhndata.cards.dataentry:0.9.0.SNAPSHOT]
	at io.uhndata.cards.serialize.ResourceToJsonAdapterFactory.lambda$leaveNode$6(ResourceToJsonAdapterFactory.java:219) [io.uhndata.cards.utils:0.9.0.SNAPSHOT]
```
More of these access denied exceptions are written every time I advance to the next form (600~ lines)

The wording/punctuation and spacing on the visit selection screen seems a little off:
![image](https://user-images.githubusercontent.com/20056751/165113095-f53f088b-f90e-4297-baaa-6a5c4fa17865.png)

Perhaps:
"To fill out pre-appointments surveys, please select one of the clinics where your upcoming appointments will take place"

There is also an extra space in the close message and the comma doesn't quite feel right

acrowthe committed:
CARDS-1716: As a patient user, I can authenticate without a token
- Minor wording changes
- Added `entered-in-error` to the list of excluded statuses

acrowthe commented:
Updated wording:
![image](https://user-images.githubusercontent.com/20056751/165131488-ef3a0351-f048-44cf-b5db-eeca057a5c70.png)

Also changed visit validation rules:

Now excluding statuses `cancelled` and `entered-in-error`, previously only excluded `cancelled`

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Removed unnecessary style rule

marta- commented:
Bug: clicking on "Terms of Use and Privacy Policy" at the bottom of the Patient identification screen when the Patient gets there without an `auth_token` results in thousands of failed calls to load the Terms of Use:

![image](https://user-images.githubusercontent.com/651980/165135862-d4374fd1-0489-4a2f-8ea7-dc589b4e6066.png)

marta- committed:
CARDS-1716: As a patient user, I can authenticate without a token
Bugfix: Clicking on "Terms of Use and Privacy Policy" at the bottom of the Patient identification screen when the Patient gets there without an auth_token results in repeated failed calls to load the Terms of Use acceptance status, and theTerms of Use are never shown

marta- commented:
> Bug: clicking on "Terms of Use and Privacy Policy" at the bottom of the Patient identification screen when the Patient gets there without an `auth_token` results in thousands of failed calls to load the Terms of Use acceptance status

Fixed in 8efe427.

sashaandjic commented:
I am unable to login as a patient (no token)

acrowthe committed:
CARDS-1716/CARDS-1735: Patient authentication cookies should expire when the token expires
- Add in expiry date for CARDS-1716 tokens

acrowthe commented:
Added in cookie expiration. Without 1735 (#1011), these tokens + expiry date will get overwritten by a token without expiry date so will not work properly yet.

The [CARDS-1716-test](https://github.com/data-team-uhn/cards/tree/CARDS-1716-test) branch has been created with 1716 and 1735, in addition to reducing the 1716 token/cookie expiration to 2 minutes for ease of testing.

To test cookies on the CARDS-1716-test branch, make a patient with MRN/HC and birth date and 1 or more valid visits (`Surveys`, future `Time`, `Location`, not `cancelled` or `entered-in-error` status and login using the PROMs tokenless authentication process. Using network inspector, you should see that the created cookie has an expiry date and once it has expired, requests should fail

![image](https://user-images.githubusercontent.com/20056751/165171949-178b5ee0-93a2-42a5-94f7-7ec5a09dac83.png)
![image](https://user-images.githubusercontent.com/20056751/165171954-b12c7f05-72f2-482c-a151-073b10f8feb6.png)
![image](https://user-images.githubusercontent.com/20056751/165171957-2d58cdf7-9215-4f5f-9fc9-12416be23f8b.png)
